synapse/util/caches/descriptors.py

```diff
@@ -46,17 +46,17 @@
 
 
 class _CachedFunction(Generic[F]):
-    invalidate = None  # type: Any
-    invalidate_all = None  # type: Any
-    prefill = None  # type: Any
-    cache = None  # type: Any
-    num_args = None  # type: Any
+    invalidate: Any = None
+    invalidate_all: Any = None
+    prefill: Any = None
+    cache: Any = None
+    num_args: Any = None
 
-    __name__ = None  # type: str
+    __name__: str
```

Comment:
Looks like the defaults of `None` are removed in a bunch of places in this file. I'm not sure if that's really correct or not?

I guess it is OK since this is a generic, not an implementation, but I find it a bit odd.

Reply:
I believe it's more correct than before, though i don't know for sure, i'm mainly afraid of the `None`s leaking through to the subclasses, becoming defaults on the class/instance variables there

Reply:
This transformation seems a bit weird in general, but it *is* a generic and I guess `None` isn't of type `str` either, so maybe it's no more nonsensical than before.

I would be careful about similar transformations elsewhere though, since it hasn't preserved the exact same semantics:

```python
>>> xyz: str
>>> xyz
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'xyz' is not defined
```

Reply:
Looks like we already do this in some other places, so you can ignore this. 😄